xds/src/main/java/io/grpc/xds/GcpAuthenticationFilter.java

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright 2021 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds;
+
+import com.google.auth.oauth2.ComputeEngineCredentials;
+import com.google.auth.oauth2.IdTokenCredentials;
+import com.google.protobuf.Any;
+import com.google.protobuf.InvalidProtocolBufferException;
+import com.google.protobuf.Message;
+import io.envoyproxy.envoy.extensions.filters.http.gcp_authn.v3.GcpAuthnFilterConfig;
+import io.envoyproxy.envoy.extensions.filters.http.gcp_authn.v3.TokenCacheConfig;
+import io.grpc.CallCredentials;
+import io.grpc.CallOptions;
+import io.grpc.Channel;
+import io.grpc.ClientCall;
+import io.grpc.ClientInterceptor;
+import io.grpc.LoadBalancer.PickSubchannelArgs;
+import io.grpc.MethodDescriptor;
+import io.grpc.auth.MoreCallCredentials;
+import io.grpc.util.LruCache;
+import io.grpc.xds.Filter.ClientInterceptorBuilder;
+import java.util.concurrent.ScheduledExecutorService;
+import javax.annotation.Nullable;
+
+/**
+ * A {@link Filter} that injects a {@link CallCredentials} to handle
+ * authentication for xDS credentials.
+ */
+public final class GcpAuthenticationFilter implements Filter, ClientInterceptorBuilder {
+
+  static final String TYPE_URL =
+      "type.googleapis.com/envoy.extensions.filters.http.gcp_authn.v3.GcpAuthnFilterConfig";
+
+  public GcpAuthenticationFilter() {
+
+  }
+
+  @Override
+  public String[] typeUrls() {
+    return new String[] { TYPE_URL };
+  }
+
+  @Override
+  public ConfigOrError<? extends FilterConfig> parseFilterConfig(Message rawProtoMessage) {
+    GcpAuthnFilterConfig gcpAuthnProto;
+    if (!(rawProtoMessage instanceof Any)) {
+      return ConfigOrError.fromError("Invalid config type: " + rawProtoMessage.getClass());
+    }
+    Any anyMessage = (Any) rawProtoMessage;
+
+    try {
+      gcpAuthnProto = anyMessage.unpack(GcpAuthnFilterConfig.class);
+    } catch (InvalidProtocolBufferException e) {
+      return ConfigOrError.fromError("Invalid proto: " + e);
+    }
+
+    // Validate cache_config
+    TokenCacheConfig cacheConfig = gcpAuthnProto.getCacheConfig();
+    if (cacheConfig != null) {
+      long cacheSize = cacheConfig.getCacheSize().getValue();
```

Comment:
Declare `long cacheSize = 10` before the `if`. Then use `cacheSize` when constructing `GcpAuthenticationConfig` instead of re-performing the null check.